File: Rectangle.Windows/src/Rectangle.Windows/Core/DragState.cs
using System;

namespace Rectangle.Windows.Core;

/// <summary>
/// 拖拽状态
/// 记录窗口拖拽过程中的状态信息
/// </summary>
public class DragState
{
    /// <summary>
    /// 是否正在拖拽
    /// </summary>
    public bool IsDragging { get; set; }
    
    /// <summary>
    /// 被拖拽的窗口句柄
    /// </summary>
    public nint DraggedWindow { get; set; }
    
    /// <summary>
    /// 拖拽开始时的鼠标位置
    /// </summary>
    public System.Drawing.Point InitialMousePos { get; set; }
    
    /// <summary>
    /// 拖拽开始时的窗口位置
    /// </summary>
    public WindowRect InitialWindowRect { get; set; }
    
    /// <summary>
    /// 拖拽开始时间
    /// </summary>
    public DateTime DragStartTime { get; set; }
    
    /// <summary>
    /// 当前鼠标位置
    /// </summary>
    public System.Drawing.Point CurrentMousePos { get; set; }
    
    /// <summary>
    /// 当前吸附区域（如果有）
    /// </summary>
    public SnapArea? CurrentSnapArea { get; set; }
    
    /// <summary>
    /// 是否已触发吸附
    /// </summary>
    public bool HasSnapped { get; set; }
    
    /// <summary>
    /// 原始窗口位置（用于 Unsnap 恢复）
    /// </summary>
    public WindowRect? OriginalRect { get; set; }
    
    /// <summary>
    /// 拖拽的按钮（左键/右键）
    /// </summary>
    public MouseButton DragButton { get; set; }

    /// <summary>
    /// 重置状态
    /// </summary>
    public void Reset()
    {
        IsDragging = false;
        DraggedWindow = 0;
        InitialMousePos = System.Drawing.Point.Empty;
        InitialWindowRect = new WindowRect();
        DragStartTime = DateTime.MinValue;
        CurrentMousePos = System.Drawing.Point.Empty;
        CurrentSnapArea = null;
        HasSnapped = false;
        OriginalRect = null;
        DragButton = MouseButton.None;
    }

    /// <summary>
    /// 计算拖拽偏移量
    /// </summary>
    public System.Drawing.Point GetDragOffset()
    {
        return new System.Drawing.Point(
            CurrentMousePos.X - InitialMousePos.X,
            CurrentMousePos.Y - InitialMousePos.Y);
    }

    /// <summary>
    /// 计算拖拽持续时间（毫秒）
    /// </summary>
    public int GetDragDurationMs()
    {
        return (int)(DateTime.Now - DragStartTime).TotalMilliseconds;
    }

    /// <summary>
    /// 计算拖拽距离（像素）
    /// </summary>
    public double GetDragDistance()
    {
        var offset = GetDragOffset();
        return Math.Sqrt(offset.X * offset.X + offset.Y * offset.Y);
    }

    /// <summary>
    /// 根据初始位置和偏移量计算新位置
    /// </summary>
    public WindowRect CalculateNewRect()
    {
        var offset = GetDragOffset();
        return new WindowRect(
            InitialWindowRect.X + offset.X,
            InitialWindowRect.Y + offset.Y,
            InitialWindowRect.Width,
            InitialWindowRect.Height);
    }
}

/// <summary>
/// 鼠标按钮
/// </summary>
public enum MouseButton
{
    None,
    Left,
    Right,
    Middle
}

/// <summary>
/// 吸附区域
/// </summary>
public class SnapArea
{
    /// <summary>
    /// 区域边界
    /// </summary>
    public System.Drawing.Rectangle Bounds { get; set; }
    
    /// <summary>
    /// 对应的窗口操作
    /// </summary>
    public WindowAction Action { get; set; }
    
    /// <summary>
    /// 区域类型
    /// </summary>
    public SnapAreaType Type { get; set; }
    
    /// <summary>
    /// 区域名称（用于日志）
    /// </summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// 检查点是否在区域内
    /// </summary>
    public bool Contains(System.Drawing.Point point)
    {
        return Bounds.Contains(point);
    }

    /// <summary>
    /// 获取区域的中心点
    /// </summary>
    public System.Drawing.Point GetCenter()
    {
        return new System.Drawing.Point(
            Bounds.Left + Bounds.Width / 2,
            Bounds.Top + Bounds.Height / 2);
    }
}

/// <summary>
/// 吸附区域类型
/// </summary>
public enum SnapAreaType
{
    /// <summary>
    /// 屏幕边缘
    /// </summary>
    Edge,
    
    /// <summary>
    /// 屏幕角落
    /// </summary>
    Corner,
    
    /// <summary>
    /// 自定义区域
    /// </summary>
    Custom
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Rust/src/core/drag_state.rs
use crate::core::rect::WindowRect;
use crate::core::action::WindowAction;

/// 鼠标按钮
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MouseButton {
    Left,
    Right,
    Middle,
}

/// 吸附区域类型
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SnapAreaType {
    Edge,
    Corner,
    Center,
}

/// 吸附区域
#[derive(Debug, Clone)]
pub struct SnapArea {
    pub action: WindowAction,
    pub area_type: SnapAreaType,
    pub name: String,
}

impl SnapArea {
    pub fn new(action: WindowAction, area_type: SnapAreaType, name: impl Into<String>) -> Self {
        Self {
            action,
            area_type,
            name: name.into(),
        }
    }
}

/// 拖拽状态
#[derive(Debug, Clone)]
pub struct DragState {
    pub is_dragging: bool,
    pub dragged_window: isize,
    pub initial_mouse_x: i32,
    pub initial_mouse_y: i32,
    pub current_mouse_x: i32,
    pub current_mouse_y: i32,
    pub initial_window_rect: WindowRect,
    pub original_rect: Option<WindowRect>,
    pub current_snap_area: Option<SnapArea>,
    pub drag_button: Option<MouseButton>,
    pub drag_start_time: std::time::Instant,
}

impl DragState {
    pub fn new() -> Self {
        Self {
            is_dragging: false,
            dragged_window: 0,
            initial_mouse_x: 0,
            initial_mouse_y: 0,
            current_mouse_x: 0,
            current_mouse_y: 0,
            initial_window_rect: WindowRect::new(0, 0, 0, 0),
            original_rect: None,
            current_snap_area: None,
            drag_button: None,
            drag_start_time: std::time::Instant::now(),
        }
    }

    /// 重置状态
    pub fn reset(&mut self) {
        self.is_dragging = false;
        self.dragged_window = 0;
        self.initial_mouse_x = 0;
        self.initial_mouse_y = 0;
        self.current_mouse_x = 0;
        self.current_mouse_y = 0;
        self.initial_window_rect = WindowRect::new(0, 0, 0, 0);
        self.original_rect = None;
        self.current_snap_area = None;
        self.drag_button = None;
    }

    /// 获取拖拽距离
    pub fn get_drag_distance(&self) -> i32 {
        let dx = self.current_mouse_x - self.initial_mouse_x;
        let dy = self.current_mouse_y - self.initial_mouse_y;
        ((dx * dx + dy * dy) as f64).sqrt() as i32
    }

    /// 获取拖拽持续时间（毫秒）
    pub fn get_drag_duration_ms(&self) -> u64 {
        self.drag_start_time.elapsed().as_millis() as u64
    }
}

impl Default for DragState {
    fn default() -> Self {
        Self::new()
    }
}
